src/JuliusSweetland.OptiKey.Core/UI/ViewModels/MainViewModel.ServiceEventHandlers.cs

```diff
@@ -2376,46 +2395,137 @@ private void SelectVoice(string voice)
+                            //commandKeyValue represents the virtual key character
+                            keyStateService.KeyDownStates[commandKeyValue].Value = KeyDownStates.Down;
+                        }
+                    }
+                    else if (commandList.First().StartsWith("KeyDown>"))
+                    {
+                        commandKeyValue = new KeyValue(commandList.First().Substring(8));
+                        Log.InfoFormat("StepListCommand: key down on {0} key", commandKeyValue.String);
+                        await keyboardOutputService.ProcessSingleKeyPress(commandKeyValue.String, KeyPressKeyValue.KeyPressType.Press);
+                        //commandKeyValue represents the virtual key character
+                        keyStateService.KeyDownStates[commandKeyValue].Value = KeyDownStates.Down;
+                    }
+                    else if (commandList.First().StartsWith("KeyUp>"))
+                    {
+                        commandKeyValue = new KeyValue(commandList.First().Substring(6));
+                        Log.InfoFormat("StepListCommand: key up on {0} key", commandKeyValue.String);
+                        await keyboardOutputService.ProcessSingleKeyPress(commandKeyValue.String, KeyPressKeyValue.KeyPressType.Release);
+                        //commandKeyValue represents the virtual key character
+                        keyStateService.KeyDownStates[commandKeyValue].Value = KeyDownStates.Up;
+                    }
+                    else if (commandList.First().StartsWith("Text>"))
+                    {
+                        commandKeyValue = new KeyValue(commandList.First().Substring(5));
+                        Log.InfoFormat("StepListCommand: text of '{0}'", commandKeyValue.String);
+                        KeySelectionResult(commandKeyValue, multiKeySelection);
                     }
-                    else if (vStep.StartsWith("Wait>"))
+                    else if (commandList.First().StartsWith("Wait>"))
                     {
-                        System.Threading.Thread.Sleep(int.Parse(vStep.Substring(5)));
+                        var waitMs = int.Parse(commandList.First().Substring(5));
+                        Log.InfoFormat("StepListCommand: wait of {0}ms", waitMs);
+                        await Task.Delay(waitMs);
                     }
                 }
+
+                if (commandList.Any())
+                {
+                    commandList.RemoveAt(0);
+                }
+                //if we are done with all commands and at the top level then the key states can be changed
+                if (!commandList.Any() && nestLevel == 0)
+                {
+                    keyStateService.KeyRunningStates[singleKeyValue].Value = false;
+                    keyStateService.KeyDownStates[singleKeyValue].Value = KeyDownStates.Up;
+                }

```

Comment:
This is the second way to end a running key. It finishes all it's tasks on it's own.

A third way to end a running key: ending it from some other key, is not yet built.